bee-tangle/src/ms_tangle.rs

```diff
@@ -0,0 +1,421 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    config::TangleConfig,
+    metadata::{IndexId, MessageMetadata},
+    solid_entry_point::SolidEntryPoint,
+    storage::StorageBackend,
+    tangle::{Hooks, Tangle, DEFAULT_CACHE_LEN},
+    urts::UrtsTipPool,
+    MessageRef,
+};
+
+use bee_message::{
+    milestone::{Milestone, MilestoneIndex},
+    Message, MessageId,
+};
+use bee_runtime::resource::ResourceHandle;
+
+use async_trait::async_trait;
+use hashbrown::HashMap;
+use log::{info, trace};
+use ref_cast::RefCast;
+use tokio::sync::Mutex;
+
+use std::{
+    ops::Deref,
+    sync::atomic::{AtomicU32, Ordering},
+};
+
+const SYNCED_THRESHOLD: u32 = 2;
+const CONFIRMED_THRESHOLD: u32 = 2;
+
+/// Tangle hooks that interoperate with Bee's storage layer.
+pub struct StorageHooks<B> {
+    #[allow(dead_code)]
+    storage: ResourceHandle<B>,
+}
+
+#[async_trait]
+impl<B: StorageBackend> Hooks<MessageMetadata> for StorageHooks<B> {
+    type Error = B::Error;
+
+    async fn get(&self, msg: &MessageId) -> Result<Option<(Message, MessageMetadata)>, Self::Error> {
+        trace!("Attempted to fetch message {:?}", msg);
+        let message = self.storage.fetch(msg).await?;
+        let meta = self.storage.fetch(msg).await?;
+        Ok(message.zip(meta))
+    }
+
+    async fn insert(&self, msg: MessageId, tx: Message, metadata: MessageMetadata) -> Result<(), Self::Error> {
```

Comment:
For this and all the others I'd use `msg_id` instead of `msg` when it's actually an id, otherwise it's a bit confusing. I'll let you decide if you want to change but at least change `tx` here as it's wrong.